feat: Support regional based timestamp

chohongm opened the pull request:
### Changelog
- Added `dateLocale` to `Constant`. It is applied to string values of message timestamp and date separator.

### Ticket
- https://sendbird.atlassian.net/browse/CLNP-3295

chohongm committed:
Support regional based timestamp

chohongm committed:
Add unit test

AhyoungRyu commented on test/utils/messageTimestamp.test.ts:
```diff
@@ -0,0 +1,31 @@
+      locale: string,
+      options?: Intl.DateTimeFormatOptions
+    ) => {
+      /**
+       * FIXME: Here is not reached but expected.
+       */
+      // console.log('## locale: ', locale);
```
Let's remove the console unless we use it :)

AhyoungRyu commented on src/utils/messageTimestamp.ts:
```diff
@@ -0,0 +1,15 @@
+    hour12: true, // ensures the time is formatted in 12-hour clock notation with AM/PM.
+  };
+  const locale = navigator.language;
+  const localTimeString: string = date.toLocaleTimeString(locale, options);
```
```suggestion
  const localTimeString = date.toLocaleTimeString(locale, options);
```
I think this should be fine without this since the return type of toLocaleTimeString is already string.

chohongm replied:
updated to
```
  const time = format(createdAt || 0, 'p', { locale });

```

AhyoungRyu reviewed (approved):
Left minor comments. LGTM

bang9 reviewed (commented):
Please double-check what the customer requested in the ticket.

For reference, uikit uses `date-fns`. We need to consider how customers can set such locale configurations for the widget in React and self-service environment.

chohongm committed:
update

chohongm committed:
Added locale to Constant

chohongm committed:
Added locale to Constant

chohongm commented:
@bang9 thanks, yes I was going to separate PRs but I changed my mind and just put all requests in this PR.

The ticket had three requests:
1. support region based `DateSeparator`
  - I opened `locale` that is injected to `SendbirdProvider`
2. support region based message timestamp.
  - I added a util function `formatCreatedAtToAMPM(createdAt, locale)` that replaced old values for `SentTime`
3. support region based message input placeholder
  - Nothing to do. they can use 
```
      stringSet={{
        MESSAGE_INPUT__PLACE_HOLDER: '메세지를 입력하세요.',
      }}
```

bang9 commented on package.json:
```diff
@@ -41,6 +42,7 @@
     "@typescript-eslint/eslint-plugin": "^5.60.1",
     "@typescript-eslint/parser": "^5.60.1",
     "@vitejs/plugin-react": "^4.2.1",
+    "date-fns": "^3.6.0",
```
https://sendbird.slack.com/archives/G01290GCDCN/p1716974520859409?thread_ts=1716974444.514649&cid=G01290GCDCN

chohongm replied:
done!

bang9 reviewed (changes requested):
type import 안된부분들이 있는거같습니다.

bang9 reviewed (changes requested):
locale 로 변경하면서, 주석/테스트 이름이 변경 안된것 같습니다.

AhyoungRyu commented on src/const.ts:
```diff
@@ -233,6 +235,10 @@ export interface Constant extends Required<ConstantProps> {
+  /**
+   * @public locale instance to be applied
+   */
+  locale: Locale;
```
If you want to keep it public, could you move it to under `autoOpen` which is the section of public props? Also we need to add this to README prop table as well.

chohongm replied:
hmm not sure about the location.
```
export interface BotStyle {
  theme: 'light' | 'dark';
  accentColor: string;
  botMessageBGColor: string;
  /** @deprecated We no longer use the autoOpen value from the dashboard. **/
  autoOpen: boolean;
}
```
autoOpen is under BotStyle. locale is unrelated. could you explain more?

chohongm committed:
updated submodule

chohongm committed:
update locale based changes

chohongm committed:
apply feeds

bang9 commented on test/utils/messageTimestamp.test.ts:
```diff
@@ -0,0 +1,16 @@
+describe('formatCreatedAtToAMPM', () => {
+  it('Verify formatCreatedAtToAMPM with mocked date.', () => {
+    const mockDate = new Date(Date.UTC(2012, 11, 20, 1, 56, 31));
+    /**
+     * Note the below expected time depends on the region in which the test is running.
+     * South Korea is 9 hours ahead of Coordinated Universal Time.
+     */
+    const expectedTime = '10:56 AM';
+    const result = formatCreatedAtToAMPM(mockDate.getTime());
+    expect(result).toBe(expectedTime);
```
이런경우 expected time string 이 실행되는 디바이스의 타임존에 따라서 바뀌기 때문에, 일관된 테스트를 보장할 수 없습니다.
따라서 아래 두가지중 하나를 테스트 하거나, 테스트를 빼거나 하면 좋을 것 같습니다.

- 정규식으로 포멧이 잘 나오는지 체크
- date-fns 의 format 함수를 mock 하여, 저희가 넘긴 파라메터로 잘 호출되는지 체크 (format 함수가 잘 동작하는지는 date-fns 의 책임이기 때문에, 리턴값을 테스트하는것은 큰 의미가 없습니다.)

chohongm replied:
흠 모킹하면 요 테스트는 format (3rd party library) 에 대한 테스트가 되는거 같아서 별 의미가 없을것 같습니다. 빼도록 할게요!

bang9 commented on src/const.ts:
```diff
@@ -157,6 +160,10 @@ export interface Constant extends Required<ConstantProps> {
+  /**
+   * @public locale instance to be applied
+   */
+  locale: Locale;
```
UIKit 에 맞춰서, locale -> dateLocale 로 모두 변경해주실 수 있을까요?

chohongm replied:
요건 저도 불확실했는데 dateLocale 로 맞추시죠~

bang9 commented on package.json:
```diff
@@ -41,6 +42,7 @@
     "@typescript-eslint/eslint-plugin": "^5.60.1",
     "@typescript-eslint/parser": "^5.60.1",
     "@vitejs/plugin-react": "^4.2.1",
+    "date-fns": "^3.6.0",
```
peerDependencies 업데이트와 peerDependenciesMeta 추가 부탁드립니다.

```json
"peerDependencies": {
  "date-fns": "^3.6.0",
  "react": "^16.8.6 || ^17.0.0 || ^18.0.0",
  "react-dom": "^16.8.6 || ^17.0.0 || ^18.0.0"
},
{
  "peerDependenciesMeta": {
    "date-fns": {
      "optional": true
    }
  }
},
```

chohongm replied:
peerDependencies 에 추가한다는건 위젯을 사용하는 고객쪽 앱에 설치가 되어야 한다는 거죠? 그럼 "date-fns": "^3.6.0", 를 차라리 dependency 에 추가하는 방법도 있지 않나요? 어떤 이유로 peerDependencies 에 추가하는 방향을 선택하셨을까요?

chohongm replied:
- 양쪽에 다른버전이 설치됐을 때의 충돌이 있을수 있음.
- 번들 사이즈를 줄일 수 있음.
요런 사유일까요?

bang9 replied:
widget 을 사용할 때, locale 을 주입하기 위해서는 date-fns 를 설치한 뒤 넣어주어야 합니다.
이때문에 peerDependencies 에 명시를 하는것이구요

하지만 locale 을 주입하지 않는다면 설치할 일도 없겠죠? 그래서 추가로 peerDependenciesMeta 로 옵셔널을 명시합니다.

(dependencies 는 옵셔널 X, 무조건 같이 설치 됨)

bang9 replied:
dependency 에 명시하지 않는 이유는 저희가 라이브러리를 빌드할 때, external 로 명시하는 패키지가 react, react-dom 밖에 없습니다. (vite.config.ts 참고)

즉, 빌드 된 코드에서 node_modules 에서 import 해서 쓸 라이브러리는 react, react-dom 두개 뿐이라는 이야기입니다.

나머지 date-fns 및 styled-components 는 빌드를 할 때 모두 소스코드에 포함시키고 있고, node_modules 에서 가져와서 사용하지 않습니다. (이때문에 현재 dependencies 로 있는 styled-components 또한 dependencies 에서 제외되어 devDependencies 로 들어가도 됩니다.)

정리하면

패키지 사용자가 저희 위젯을 설치할때는 react 와 react-dom 만 있으면 저희 라이브러리를 사용할 수 있고 (이미 date-fns, styled-components 같은 코드들이 라이브러리 코드에 포함이 되어있기 때문에)
추가로 locale 을 지정하려면 date-fns 를 설치하여 locale 만 넘겨주면 됩니다.

chohongm replied:
자세한 설명 감사합니다 @bang9

chohongm committed:
apply feeds

chohongm committed:
add packages/uikit

bang9 reviewed (approved):
lgtm!
